learntools/core/asserts.py

```diff
@@ -1,31 +1,107 @@
+                    assert_fn.__name__))
+            kwargs['name'] = '`{}`'.format(var)
+        return assert_fn(*args, **kwargs)
+    return wrapped
+
+@name_or_var
+def assert_equal(actual, expected, name, failure_factory=None):
+    """Assert a protean notion of equality specific to the use case of learntools
+    checking. Subclasses of EqualityCheckProblem ultimately use this function
+    in their check method.
+
+    Includes special cases for several types of expected values, including Pandas
+    objects, ndarrays, and floats.
+    """
+    # We default to == comparison, but have special cases for certain data types.
+    if isinstance(expected, float):
+        assert isinstance(actual, numbers.Number), \
+            "Expected {} to be a number, but had value `{!r}` (type = `{}`)".format(
+                name, actual, type(actual).__name__)
+        check = math.isclose(actual, expected, rel_tol=1e-06)
+    elif isinstance(expected, pd.DataFrame):
+        assert_df_equals(actual, expected, name)
+        return
+    elif isinstance(expected, pd.Series):
+        assert_series_equals(actual, expected, name)
+        return
+    elif isinstance(actual, np.ndarray) or isinstance(expected, np.ndarray):
+        check = np.array_equal(actual, expected)
+    else:
+        check = actual == expected
+    if failure_factory:
+        # This optional kwarg lets the caller pass a function to generate a custom
+        # failure message. Currently only used in the Python ex1 favourite color question.
+        failure_message = failure_factory(name, actual, expected)
+    else:
+        failure_message = "Incorrect value for {}: `{}`".format(
+                name, repr(actual))
+    assert check, failure_message
 
-def assert_has_columns(df, cols, name=None, strict=False):
-    df_desc = "dataframe"
-    if name:
-        df_desc += ' `{}`'.format(name)
+@name_or_var
+def assert_has_columns(df, cols, name="dataframe", strict=False):
+    """Assert that the given dataframe contains columns with the given names.
+    If strict is True, then assert it has *only* those columns.
+    """
     for col in cols:
         assert col in df.columns, "Expected {} to have column `{}`".format(
-                df_desc, col
+                name, col
                 )
     if strict:
         for col in df.columns:
-            msg = "Unexpected column in {}: `{}`".format(df_desc, col)
+            msg = "Unexpected column in {}: `{}`".format(name, col)
             assert col in cols, msg
 
-def assert_isinstance(cls, **named_things):
-    # TODO: Kind of a bad idea to use kwargs here, because names might not be
-    # valid Python identifiers in some cases.
-    for name, val in named_things.items():
-        assert isinstance(val, cls), "Expected {} to have type `{!r}` but had type `{!r}`".format(name, cls, type(val))
+@name_or_var
```

Comment:
It's interesting to see some of these assert helpers now, after having written the manual equivalent inside many check() methods.

Do you have an opinion about whether, moving forward, it's worth any effort to remember `assert_hascolumns` and `assert_isinstance` and use these, vs doing it manually when writing check methods?

On the face of it, these methods seem to be doing something sufficiently simple that I don't feel bad that I'm likely to forget they exist.  But could be missing something.

Reply:
> Do you have an opinion about whether, moving forward, it's worth any effort to remember assert_hascolumns and assert_isinstance and use these, vs doing it manually when writing check methods?

Yes, I think it is worth it. One big advantage of using them for even simple checks is that they take care of producing a reasonable failure message, including formatting. When writing them out by hand, it's easy to slip in little formatting errors, e.g. forgetting backticks.

My process when implementing some kind of `.check` validation is to check whether it already exists in `asserts.py`, and if it doesn't, I usually just add it there, since I figure I'll probably use it again later (unless it's super specific).

I was actually thinking of possibly spending a bit of time today refactoring a few exercise modules to use assertion helpers.

Reply:
For better or for worse, I expect we'll do some copy/pasting of existing modules to start new ones in the future.

So if you grep to find the locations of all the `assert type(` and replace them, I think it's likely this will remind us on a regular basis about these helper functions, which may then help us develop the workflow you describe